eureka2-write-server/src/main/java/com/netflix/eureka2/server/service/bootstrap/PeerRegistryBootstrapService.java

```diff
@@ -0,0 +1,118 @@
+package com.netflix.eureka2.server.service.bootstrap;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.LinkedHashSet;
+
+import com.netflix.eureka2.Server;
+import com.netflix.eureka2.interests.ChangeNotification;
+import com.netflix.eureka2.interests.ChangeNotifications;
+import com.netflix.eureka2.interests.Interests;
+import com.netflix.eureka2.registry.Source;
+import com.netflix.eureka2.registry.SourcedEurekaRegistry;
+import com.netflix.eureka2.registry.instance.InstanceInfo;
+import com.netflix.eureka2.server.resolver.EurekaEndpoint;
+import com.netflix.eureka2.server.resolver.EurekaEndpointResolver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Notification;
+import rx.Observable;
+import rx.Scheduler;
+import rx.functions.Func1;
+import rx.schedulers.Schedulers;
+
+/**
+ * @author Tomasz Bak
+ */
+@Singleton
+public class PeerRegistryBootstrapService implements RegistryBootstrapService {
+
+    private static final Logger logger = LoggerFactory.getLogger(PeerRegistryBootstrapService.class);
+
+    private final EurekaEndpointResolver bootstrapResolver;
+    private final Scheduler scheduler;
+
+    @Inject
+    public PeerRegistryBootstrapService(PeerBootstrapResolverProvider bootstrapResolverProvider) {
+        this(bootstrapResolverProvider.get(), Schedulers.computation());
+    }
+
+    public PeerRegistryBootstrapService(EurekaEndpointResolver bootstrapResolver, Scheduler scheduler) {
+        this.bootstrapResolver = bootstrapResolver;
+        this.scheduler = scheduler;
+    }
+
+    @Override
+    public Observable<Void> loadIntoRegistry(final SourcedEurekaRegistry<InstanceInfo> registry, final Source source) {
+        return bootstrapResolver.eurekaEndpoints()
+                .compose(ChangeNotifications.<EurekaEndpoint>buffers())
+                .compose(ChangeNotifications.<EurekaEndpoint>snapshots())
+                .take(1)
+                .flatMap(new Func1<LinkedHashSet<EurekaEndpoint>, Observable<EurekaEndpoint>>() {
+                    @Override
+                    public Observable<EurekaEndpoint> call(LinkedHashSet<EurekaEndpoint> eurekaEndpoints) {
+                        return Observable.from(eurekaEndpoints);
+                    }
+                }).flatMap(new Func1<EurekaEndpoint, Observable<Boolean>>() {
+                    @Override
+                    public Observable<Boolean> call(EurekaEndpoint endpoint) {
+                        return loadRegistryFrom(new Server(endpoint.getHostname(), endpoint.getInterestPort()), registry, source);
+                    }
+                })
+                .filter(new Func1<Boolean, Boolean>() { // true - download succeeded, false - failed, so try another server
```

Comment:
Bootstrap status can be true or false. We have here a filter that should terminate as soon as bootstrap == true is received. This implementation is however flawed. I have reimplemented it.